src/redisearch_rs/result_processor/src/ffi.rs

```diff
@@ -0,0 +1,236 @@
+/// must be the first member of the [`ResultProcessor`] struct in order to guarantee that we can cast a `*mut ResultProcessor<P>`
+/// to a `*mut Header` (which is what the C code expects to receive).
+///
+/// # Pinning Safety Contract
+///
+/// Result processors intrusively linked, where one result processor has the pointer to the
+/// previous (forming an intrusively singly-linked list). This places a big safety invariant on all code
+/// that touches this data structure: Result processors must *never* be moved while part of a QueryIterator
+/// chain. Accidentally moving a processor will remove in a broken link and undefined behaviour.
+///
+/// This invariant is implicit in the C code where its uncommon to move the heap allocated objects (which would
+/// mean memcopying them around). Rust move semantics make it very easy to break this invariant however, which is why
+/// all APIs will only ever return either immutable or pinned mutable references to the data structure.
+///
+/// This crate has to ensure that these pinning invariants are upheld internally, but when sending pointers through the FFI boundary
+/// this is impossible to guarantee. Thankfully, as mentioned above its quite rare to move out of pointers, so its reasonably safe.
+#[repr(C)]
+#[derive(Debug)]
+pub struct Header {
+    /// Reference to the parent structure
+    pub(super) parent: *mut c_void, // QueryIterator*
+
+    /// Previous result processor in the chain
+    pub(super) upstream: *mut Self,
+
+    /// Type of result processor
+    ty: ResultProcessorType,
+
+    /// time measurements
+    giltime: timespec,
+
+    /// "VTable" function. Pulls SearchResults out of this result processor.
+    pub(super) next: Option<ResultProcessorNext>,
+    /// "VTable" function. Destroys and deallocates the result processor. This is only ever called by C code.
+    free: Option<ResultProcessorFree>,
+
+    // ResultProcessors *must* be !Unpin to ensure they are not moved about breaking the linked list.
+    _unpin: PhantomPinned,
+}
+
+pin_project! {
+    #[derive(Debug)]
+    #[repr(C)]
+    pub struct ResultProcessor<P> {
+        header: Header,
+        result_processor: P,
+    }
+}
+
+impl<P> ResultProcessor<P>
+where
+    P: crate::ResultProcessor,
+{
+    /// Construct a new FFI-ResultProcessor from the provided [`crate::ResultProcessor`] implementer.
+    pub fn new(result_processor: P) -> Self {
+        Self {
+            header: Header {
+                parent: ptr::null_mut(),
+                upstream: ptr::null_mut(),
+                ty: P::TYPE,
+                giltime: timespec {
+                    tv_sec: 0,
+                    tv_nsec: 0,
+                },
+                next: Some(Self::result_processor_next),
+                free: Some(Self::result_processor_free),
+                _unpin: PhantomPinned,
+            },
+            result_processor,
```

Comment:
Let's explain why it's OK for `parent` and `upstream` to be set to a null pointer.